pkg/util/encode/encode.go

```diff
@@ -42,17 +46,19 @@ func Encode(obj runtime.Object, format string) ([]byte, error) {
 // EncodeTo converts the provided object to the specified format and
 // writes the encoded data to the provided io.Writer.
 func EncodeTo(obj runtime.Object, format string, w io.Writer) error {
-	encoder, err := EncoderFor(format)
+	encoder, err := EncoderFor(format, obj.GetObjectKind().GroupVersionKind().GroupVersion())
 	if err != nil {
 		return err
 	}
-
 	return errors.WithStack(encoder.Encode(obj, w))
 }
 
 // EncoderFor gets the appropriate encoder for the specified format.
-func EncoderFor(format string) (runtime.Encoder, error) {
+func EncoderFor(format string, gv schema.GroupVersion) (runtime.Encoder, error) {
```

Comment:
I need to look at this in a little more detail.

Reply:
Alright, I wrapped my head around this - let me write up some notes.

Reply:
So - I *think* our primary goal here is that, when we serialize objects to JSON/YAML and output them, via `velero backup get`, `velero install`, etc., we need to have the `Kind` and `APIVersion` fields (i.e. the `TypeMeta`) filled in.  Specifically for `velero install`, this enables the output to be `kubectl apply`-ed.

`encode.go` today only knows about the Velero types. So, if it's given a structured object that's a Velero type, it will encode it and include the `TypeMeta`. If it's given an unstructured object without its `TypeMeta` filled it, it will encode it, but won't add the `TypeMeta`. And if it's given a structured object that's not a Velero type (e.g. a `Deployment`), it won't encode it.

One option here is to enhance `encode.go` to know about other non-Velero types, e.g. the Kube types that we care about. We'd do this along the lines of what you have already in this PR, though there are a couple things that should change. If we do this, then we can encode all objects that we care about, without issue.  The drawback to this option is that, since you're outputting *unstructured* objects during `velero install`, the encoder can't help us by filling in the `TypeMeta` -- we always have to set it explicitly ourselves when creating the unstructured objects.

The other option is to switch to outputting structured objects. We'd have to modify `output.go` -- my initial thought is that we'd switch from:
```go
func PrintWithFormat(c *cobra.Command, obj runtime.Object) (bool, error)
```
to
```go
func PrintWithFormat(c *cobra.Command, objs ...runtime.Object) (bool, error)
```

That way, we could just pass a list of structured objects to it to output. It would have to be enhanced to add separators as appropriate so the overall output could still be `kubectl apply`-ed.  The benefit of this approach is that we don't have to explicitly set `TypeMeta` anywhere - if we use the structured representations, and we set up the `runtime.Scheme` properly in `encode.go`, then it'll automatically get filled in during encoding.

My gut says the second option is more ideal, though refining the first one is probably less work at the moment. WDYT?

FYI here's my hacked-up modification to `encode.go`, that supports both encoding unstructured objects with their `TypeMeta` filled in, **and** encoding Velero & Kube structured objects without `TypeMeta` filled in:
```go
// EncoderFor gets the appropriate encoder for the specified format.
func EncoderFor(format string, obj runtime.Object) (runtime.Encoder, error) {
	scheme := runtime.NewScheme()
	vscheme.AddToScheme(scheme) // vscheme "github.com/heptio/velero/pkg/generated/clientset/versioned/scheme"
	kscheme.AddToScheme(scheme) // kscheme "k8s.io/client-go/kubernetes/scheme"

	codecs := serializer.NewCodecFactory(scheme)

	var encoder runtime.Encoder
	desiredMediaType := fmt.Sprintf("application/%s", format)
	serializerInfo, found := runtime.SerializerInfoForMediaType(codecs.SupportedMediaTypes(), desiredMediaType)
	if !found {
		return nil, errors.Errorf("unable to locate an encoder for %q", desiredMediaType)
	}
	if serializerInfo.PrettySerializer != nil {
		encoder = serializerInfo.PrettySerializer
	} else {
		encoder = serializerInfo.Serializer
	}

	
	if gvks, _, _ := scheme.ObjectKinds(obj); len(gvks) > 0 {
		return codecs.EncoderForVersion(encoder, gvks[0].GroupVersion()), nil
	}

	return encoder, nil
}
```

Reply:
I would prefer to use the structured types, if possible. What exists now was trying to find a way to pass a single `UnstructuredList` (since nothing else besides `[]runtime.Object` could hold everything` in and have the current printing machinery 'just work'.

I think the structured approach allows the whole of the install code base to use `[]runtime.Object` instead of the `UnstructuredList`, which is also a win. That means that I'll have to redo some of the work we have right now, but longer term it's more maintainable as we'll get more compiler help.

I'll double check if the TypeMeta fields for anything in the `pkg/install` package are used - I think it makes sense to leave those populated (for things like filtering/grouping), but not to add TypeMeta to all the other, existing create commands.

Reply:
I've got `PrintWithFormat` modified to take the variadic list, but I'll need to think some on how to properly print out the installed resources as JSON. With the in-tree types it's fine, since they're in a list type. Printing things one at a time is fine in YAML, since they can be broken up with a single line. JSON needs to wrap the whole list as an array, and can't have a comma on the last item.

I'll do more research to see if we can manage to do this without making it too gross.

Reply:
How important is omitting the empty `status` fields? It looks like by using the structured types we're going to have those output. This is true of our current create commands with our own data types.

Reply:
Ah, yeah, having `status` be output is annoying.  Does it cause issues with `kubectl apply`-ing the resources? Or is it just a UX thing? Let me think about this some more..

Reply:
Looking back at my notes, it does for CustomResourceDefinitions. One of the status fields comes in as null and causes an error. I believe those are the only resources that require removing it, the rest is cosmetic.

Reply:
Hmm. I'm thinking maybe we shouldn't try to switch to non-unstructured objects right now - we can get this PR merged more or less as-is (I'll add a few specific comments to `encode.go`) - and then look at switching afterwards.

Reply:
That seems reasonable to me. I think switching `EncoderFor` to use the logic you've provided here is probably a better fix for our other create commands than populating them all with TypeMeta manually.